src/libexpr/eval.cc

```diff
@@ -199,6 +191,10 @@ void initGC()
     /* Initialise the Boehm garbage collector. */
     GC_set_all_interior_pointers(0);
 
+    /* Values are using interrior pointers to pack the value type. */
+    GC_register_displacement(1);
+    GC_register_displacement(2);
```

Comment:
From what I understand from the documentation, `GC_set_all_interior_pointers(0)` is used to tell the GC that we do not have any random pointer into a GC allocation.

The `GC_register_displacement` is used for for known interior pointers.  I have not read anything from the documentation which suggests that they cannot be used together.

I tried this patched version of nix to generate a dot graph with `nixos/maintainers/option-usages.nix` and I did not had any OOM after running 45 minutes at a steady usage of 1GB.